lib/handler/connect.c

```diff
@@ -967,8 +1027,69 @@ static int on_req(h2o_handler_t *_handler, h2o_req_t *req)
     return 0;
 }
 
-void h2o_connect_register(h2o_pathconf_t *pathconf, h2o_proxy_config_vars_t *config, h2o_connect_acl_entry_t *acl_entries,
-                          size_t num_acl_entries)
+static int on_req_classic_connect(h2o_handler_t *_handler, h2o_req_t *req)
+{
+    struct st_connect_handler_t *handler = (void *)_handler;
+    h2o_iovec_t host;
+    uint16_t port;
+    int is_tcp;
+
+    if (req->upgrade.base != NULL) {
+        return -1;
+    } else if (h2o_memis(req->method.base, req->method.len, H2O_STRLIT("CONNECT"))) {
+        /* old-style CONNECT */
+        is_tcp = 1;
+    } else if (h2o_memis(req->method.base, req->method.len, H2O_STRLIT("CONNECT-UDP"))) {
+        /* masque (draft 03); host and port are stored the same way as ordinary CONNECT
+         * TODO remove code once we drop support for draft-03 */
+        if (!handler->config.support_masque_draft_03) {
+            h2o_send_error_405(req, "Method Not Allowed", "Method Not Allowed", H2O_SEND_ERROR_KEEP_HEADERS);
+            return 0;
+        }
+        is_tcp = 0;
+    } else {
+        /* it is not the task of this handler to handle non-CONNECT requests */
+        return -1;
+    }
+
+    /* parse host and port from authority, unless it is handled above in the case of extended connect */
+    if (h2o_url_parse_hostport(req->authority.base, req->authority.len, &host, &port) == NULL || port == 0 || port == 65535) {
+        h2o_send_error_400(req, "Bad Request", "Bad Request", H2O_SEND_ERROR_KEEP_HEADERS);
+        return 0;
+    }
+
+    return on_req_core((void *)handler, req, host, port, is_tcp, 1);
+}
+
+/**
+ * handles RFC9298 requests
+ */
+static int on_req_connect_udp(h2o_handler_t *handler, h2o_req_t *req)
+{
+    h2o_iovec_t host;
+    uint16_t port;
+
+    /* reject requests wo. upgrade: connect-udp */
+    if (!(req->upgrade.base != NULL && h2o_lcstris(req->upgrade.base, req->upgrade.len, H2O_STRLIT("connect-udp"))))
+        return -1;
+
+    /* check method */
+    if (!(req->version < 0x200 ? h2o_memis(req->method.base, req->method.len, H2O_STRLIT("GET"))
+                               : h2o_memis(req->method.base, req->method.len, H2O_STRLIT("CONNECT"))))
+        return -1;
+
+    /* masque (RFC 9298); parse host/port */
+    if (!masque_decode_hostport(&req->pool, req->path_normalized.base + req->pathconf->path.len,
+                                req->path_normalized.len - req->pathconf->path.len, &host, &port)) {
+        h2o_send_error_400(req, "Bad Request", "Bad Request", H2O_SEND_ERROR_KEEP_HEADERS);
```

Comment:
Would it be possible to hook this error reply into `record_error` and `send_connect_error` in order to improve debuggability?